packages/harmony/src/components/input/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,178 @@
+  const showMaxlengthWarningColor =
+    maxLength && characterCount >= maxLengthWarningThreshold * maxLength
+
+  // Styles for the root of the input
+  const inputRootStyle = {
+    [styles.default]: size === TextInputSize.DEFAULT,
+    [styles.small]: size === TextInputSize.SMALL,
+    [styles.warning]: warningProp,
+    [styles.error]: error,
+    [styles.focused]: isFocused,
+    [styles.disabled]: disabled,
+    [styles.required]: required
+  }
+
+  // Styles for the input element itself
+  const inputElStyle = {
+    [styles.default]: size === TextInputSize.DEFAULT,
+    [styles.small]: size === TextInputSize.SMALL,
+    [styles.disabled]: disabled
+  }
+
+  const inputRender = (
+    <Flex
+      className={cn(styles.inputAdornmentRow, { [styles.focused]: isFocused })}
+    >
+      {startAdornmentText && shouldShowAdornments ? (
+        <Text variant='label' size='l' color='subdued'>
+          {startAdornmentText}
+        </Text>
+      ) : null}
+      <input
+        onFocus={handleFocus}
+        onBlur={handleBlur}
+        ref={ref}
+        className={cn(styles.inputElement, inputClassName, inputElStyle)}
+        value={value}
+        maxLength={maxLength}
+        disabled={disabled}
+        placeholder={shouldShowPlaceholder ? placeholder : undefined}
+        aria-label={labelText ?? props['aria-label']}
+        aria-required={required}
+        id={id}
+        {...other}
+      />
+      {endAdornmentText && shouldShowAdornments ? (
+        <Text variant='label' size='l' color='subdued'>
+          {endAdornmentText}
+        </Text>
+      ) : null}
+    </Flex>
+  )
+
+  return (
+    <Flex className={cn(styles.root, className)} direction='column'>
+      <label
+        htmlFor={id}
+        className={cn(styles.inputRoot, inputRootClassName, inputRootStyle)}
+      >
+        {StartIcon ? <StartIcon /> : null}
+        <div className={styles.innerRootContainer}>
+          {shouldShowLabel ? (
+            <Flex
+              className={styles.labelRow}
+              direction='row'
+              alignItems='center'
+              justifyContent='space-between'
+              gap='s'
+            >
+              <Text
+                variant='body'
+                tag='span'
+                size='l'
+                className={cn(styles.label, {
+                  [styles.hasValue]: hasValue
+                })}
+              >
+                {labelText}
+              </Text>
+              {shouldShowMaxLengthText ? (
+                <Text
+                  variant='body'
+                  size='xs'
+                  tag='span'
+                  color={showMaxlengthWarningColor ? 'warning' : 'default'}
+                >
+                  {characterCount}/{maxLength}
+                </Text>
+              ) : null}
+            </Flex>
+          ) : null}
+          {inputRender}
+        </div>
+
+        {EndIcon ? <EndIcon /> : null}
+        {children}
+      </label>
+      {helperText ? (
```

Comment:
This feels like _maybe_ it should be left to the caller? I tend to lean towards composability vs coupling this sort of thing. In this case, maybe a render fn with `hasError` and the caller can define the text component?

The way it is here is convenient as a dev as they'll likely be coupled a lot, so I can see it both ways, but I often find limiting this sort of thing to a string bites us sometimes when we want to add eg icons or whatever else in a future special case.

This might make sense to be the children prop, too, and have that children prop be a render fn with the error state.

Reply:
Imo, I don't see much benefit of making it more composable here. If this piece was composable the consumer would have to handle color, size, and gap spacing which all come from the harmony figmas. Plus, I think the majority of cases will be wanting the "standard behavior" anyways; and if it's non-standard we most likely have some room to push back on designs or we can revisit this implementation then.
The other reason is that there's technically already an "escape hatch". If no `helperText` is provided, it's not rendered so if a special case does arise, the composer still has a way to manage it. Honestly this already feels like the best of both worlds; we get a quick & easy access for the majority of cases and an DIY option for special cases.

Also, fwiw, I personally lean somewhere in the middle between composability vs abstracted; maybe a bit more biased towards abstracted. My opinion with design system components is that they should "quickly solve for standard use cases, but not restrict special cases (ideally)". 
With the design system at Bestow, I actually started out with components more heavily favoring composability and found that most of the time it was unnecessary extra effort; but at worst created refactoring headaches if things changed on the design system side. 
Also I generally find it easier to start more abstracted and peel back layers as situations arise. Also in a future state when the majority of designs are "harmonized" there's less reasons to deviate from the standard stuff anyways

Happy to discuss further though

Reply:
SGTM! Was a super nitpick. A happy medium here might be typing the prop as a `ReactNode` and leaving everything else, but honestly I'm fine with how you have it and we can always revisit!

Reply:
Oh yeah I actually like doing that regardless tbh. I updated the type